src/app/rpmostree-override-builtins.cxx

```diff
@@ -89,7 +93,63 @@ handle_override (RPMOSTreeSysroot  *sysroot_proxy,
   g_autoptr(GVariant) options = g_variant_ref_sink (g_variant_dict_end (&dict));
 
   g_autoptr(GVariant) previous_deployment = rpmostree_os_dup_default_deployment (os_proxy);
+  g_autoptr(GPtrArray) override_replace_final = g_ptr_array_new();
+  GUnixFDList fds;
+  g_autoptr(GUnixFDList) list_fds;
 
+  if (override_replace && opt_ex_pin_from_repos)
+    {
+      g_autoptr(GPtrArray) queries = g_ptr_array_new();
+
+      for (const char *const* it = override_replace; it && *it; it++)
+        {
+          auto pkg = *it;
+
+          if(rpmostreecxx::is_http_arg(pkg) || rpmostreecxx::is_rpm_arg(pkg))
+            {
+              g_ptr_array_add(override_replace_final, (gpointer)pkg);
+              continue;
+	    }
+          else
+            {
+              g_ptr_array_add(queries, (gpointer)pkg);
+            }
+        }
+      if (queries->len > 0)
+        {
+          g_ptr_array_add(queries, NULL);
+          g_autofree char *refresh_transaction_address = NULL;
+
+          if (!rpmostree_os_call_refresh_md_sync(os_proxy, options, &refresh_transaction_address,
+                                                 cancellable, error))
+            return FALSE;
+          if (!rpmostree_transaction_get_response_sync(sysroot_proxy, (const char*)refresh_transaction_address,
+                                                       cancellable, error))
+            return FALSE;
+          g_autoptr(GVariant) fd_idxs = NULL;
+          if (!rpmostree_osexperimental_call_download_packages_sync (osexperimental_proxy,
+                                                                     (const char *const*) queries->pdata,
+                                                                     &fds, &fd_idxs, &list_fds,
+                                                                     cancellable, error))
+            return FALSE;
+
+          const int nfds = g_unix_fd_list_get_length (list_fds);
+          const guint n = g_variant_n_children (fd_idxs);
+          gint *new_fds = g_new0 (gint, n+1);
+
+          for (guint i = 0; i < n; i++)
+            {
+              g_autoptr(GVariant) hv = g_variant_get_child_value (fd_idxs, i);
+              g_assert (hv);
+              gint32 h = g_variant_get_handle (hv);
+              g_assert (0 <= h && h < nfds);
+              new_fds[i] = fds[h];
+            }
+
+          new_fds[n] = -1;
+          override_replace = (const char *const*) new_fds;
```

Comment:
This will not work, because it needs to be an array of paths, not fds.

Though hmm, this does raise an interesting idea. Instead of having to plumb through the fds all the way from here to the sort function, we could instead convert the fds into filepaths using `glnx_fdrel_abspath`. It won't end in `.rpm` though, but I think we can (should anyway) beef up `is_rpm_arg()` to support `file://` too so it's still possible to pass local RPMs without having to rename them.

Does that make sense?

Reply:
Yes, I think it does

Reply:
In that case, the download_packages API would pass out a list of filepaths ? instead of a list of fds? Or the client receives fds and is the one translating them?